extension/dynamicconfig/service/file/backend.go

```diff
@@ -0,0 +1,114 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	"go.opentelemetry.io/collector/config"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/dynamicconfig/model"
+	pb "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/dynamicconfig/v1"
+	res "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+)
+
+// file.Backend is a ConfigBackend that uses a local file to determine
+// what schedules to change. The file is read live, so changes to it will
+// reflect immediately in the configs.
+type Backend struct {
+	viper *viper.Viper
+
+	mu          sync.Mutex
```

Comment:
It would be nice to document what are protected by the mutex. What's better, is there golang equivalent of C++ Thread safety analysis: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

Reply:
Great point, I read there was a convention that the fields protected by a mutex should come directly below it -- which is why there's the layout pattern for the two `backend.go` files. I've also added comments to clarify the protected fields.

There doesn't seem to be an equivalent to the static thread safety analysis extension available from the clang compiler, but go does offer thread-safety checking at runtime when a binary is compiled or tested using the `-race` flag. All tests have been executed with `-race` to check for thread safety issues, and seem to be all right.